File: packages/php/src/Seguranca/Ataques/AtaqueChaveRelacionada.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Ataque de chave relacionada: chaves que diferem por um padrão fixo (1 bit,
 * 0xFF, complemento) não devem gerar keystreams correlacionados. Um key
 * schedule fraco faz `keystream(key)` e `keystream(key ^ delta)` compartilharem
 * estrutura (distância de Hamming baixa).
 */
class AtaqueChaveRelacionada implements AtaqueInterface
{
    public function __construct(
        private int $tamanhoBloco = 32,
        private int $ivPorDelta = 3,
        private float $limiteMedia = 0.45,
        private float $limitePior = 0.3,
    ) {
    }

    public function nome(): string
    {
        return 'Chave relacionada (related-key)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $chaveBase = $alvo->chaveDeTeste();
        $len = strlen($chaveBase);

        $primeiro = $alvo->gerarKeystreamBruto($chaveBase, random_bytes($alvo->tamanhoIv()), 'enc', $this->tamanhoBloco);
        if ($primeiro === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $deltas = [];
        for ($i = 0; $i < $len; $i++) {
            $d = str_repeat("\x00", $len);
            $d[$i] = "\x01";
            $deltas[] = $d;
        }
        for ($i = 0; $i < $len; $i++) {
            $d = str_repeat("\x00", $len);
            $d[$i] = "\xFF";
            $deltas[] = $d;
        }
        $complemento = '';
        for ($i = 0; $i < $len; $i++) {
            $complemento .= chr(ord($chaveBase[$i]) ^ 0xFF);
        }
        $deltas[] = $complemento;

        $valores = [];
        $totalBits = $this->tamanhoBloco * 8;

        foreach ($deltas as $delta) {
            $chaveRelacionada = $chaveBase;
            for ($i = 0; $i < $len; $i++) {
                $chaveRelacionada[$i] = chr(ord($chaveRelacionada[$i]) ^ ord($delta[$i]));
            }
            for ($s = 0; $s < $this->ivPorDelta; $s++) {
                $iv = random_bytes($alvo->tamanhoIv());
                $ks1 = $alvo->gerarKeystreamBruto($chaveBase, $iv, 'enc', $this->tamanhoBloco);
                $ks2 = $alvo->gerarKeystreamBruto($chaveRelacionada, $iv, 'enc', $this->tamanhoBloco);
                $valores[] = $this->bitsDiferentes($ks1, $ks2) / $totalBits;
            }
        }

        $media = array_sum($valores) / count($valores);
        $pior = min($valores);
        $vulneravel = $media < $this->limiteMedia || $pior < $this->limitePior;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'alta' : 'info',
            detalhes: sprintf(
                'média=%.1f%%, pior=%.1f%% em %d pares (limites: média>=%.0f%%, pior>=%.0f%%)',
                $media * 100,
                $pior * 100,
                count($valores),
                $this->limiteMedia * 100,
                $this->limitePior * 100,
            ),
            dados: ['media' => $media, 'pior' => $pior],
        );
    }

    private function bitsDiferentes(string $a, string $b): int
    {
        $diff = 0;
        $len = min(strlen($a), strlen($b));
        for ($i = 0; $i < $len; $i++) {
            $diff += substr_count(decbin(ord($a[$i]) ^ ord($b[$i])), '1');
        }
        return $diff;
    }
}

File: packages/dart/lib/Seguranca/Ataques/AtaqueChaveRelacionada.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Ataque de chave relacionada: chaves que diferem por um padrão fixo (1 bit,
/// 0xFF, complemento) não devem gerar keystreams correlacionados. Um key
/// schedule fraco faz `keystream(key)` e `keystream(key ^ delta)` compartilharem
/// estrutura (distância de Hamming baixa).
class AtaqueChaveRelacionada extends Ataque {
  final int _tamanhoBloco;
  final int _ivPorDelta;
  final double _limiteMedia;
  final double _limitePior;

  AtaqueChaveRelacionada({
    int tamanhoBloco = 32,
    int ivPorDelta = 3,
    double limiteMedia = 0.45,
    double limitePior = 0.3,
  }) : _tamanhoBloco = tamanhoBloco,
       _ivPorDelta = ivPorDelta,
       _limiteMedia = limiteMedia,
       _limitePior = limitePior;

  @override
  String nome() => 'Chave relacionada (related-key)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chaveBase = toBytes(alvo.chaveDeTeste());
    final len = chaveBase.length;

    final deltas = <Uint8List>[];
    for (var i = 0; i < len; i++) {
      final d = Uint8List(len);
      d[i] = 0x01;
      deltas.add(d);
    }
    for (var i = 0; i < len; i++) {
      final d = Uint8List(len);
      d[i] = 0xff;
      deltas.add(d);
    }
    final complemento = Uint8List.fromList(
      chaveBase.map((b) => b ^ 0xff).toList(),
    );
    deltas.add(complemento);

    final valores = <double>[];
    final totalBits = _tamanhoBloco * 8;

    for (final delta in deltas) {
      final chaveRelacionada = Uint8List.fromList(chaveBase);
      for (var i = 0; i < len; i++) {
        chaveRelacionada[i] ^= delta[i];
      }
      for (var s = 0; s < _ivPorDelta; s++) {
        final iv = randomBytes(alvo.tamanhoIv());
        final ks1 = alvo.gerarKeystreamBruto(
          chaveBase,
          iv,
          'enc',
          _tamanhoBloco,
        );
        final ks2 = alvo.gerarKeystreamBruto(
          chaveRelacionada,
          iv,
          'enc',
          _tamanhoBloco,
        );
        valores.add(bitsDiferentes(ks1, ks2) / totalBits);
      }
    }

    final media = valores.reduce((a, b) => a + b) / valores.length;
    var pior = valores[0];
    for (final v in valores) {
      if (v < pior) pior = v;
    }
    final vulneravel = media < _limiteMedia || pior < _limitePior;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.alta : Severidade.info,
      'média=${(media * 100).toStringAsFixed(1)}%, '
          'pior=${(pior * 100).toStringAsFixed(1)}% '
          'em ${valores.length} pares (limites: '
          'média>=${(_limiteMedia * 100).toStringAsFixed(0)}%, '
          'pior>=${(_limitePior * 100).toStringAsFixed(0)}%)',
      {'media': media, 'pior': pior},
    );
  }
}
